File: flutter/flutter-animated_barnacle/lib/User/UserService.dart
import 'package:animated_barnacle/User/UserModel.dart';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';

class UserService {
  late User user;
  late List<User> users;
  final String baseUrl = "http://127.0.0.1:4001/users";

  void showErrorDialog(String message, context) {
      showDialog(
        context: context,
        builder: (BuildContext context) {
          return AlertDialog(
            title: Text("Error"),
            content: Text(message),
            actions: [
              TextButton(
                onPressed: () {
                  Navigator.of(context).pop();
                },
                child: Text("OK"),
              ),
            ],
          );
        },
      );
    }

      void showSuccessDialog(String message, context) {
      showDialog(
        context: context,
        builder: (BuildContext context) {
          return AlertDialog(
            title: Text("Success"),
            content: Text(message),
            actions: [
              TextButton(
                onPressed: () {
                  Navigator.of(context).pop();
                  // Navigate to the next page if login is successful
                },
                child: Text("OK"),
              ),
            ],
          );
        },
      );
    }

  Future<bool> createUser(String username, String email,  String password) async {

        final response = await http.post(
          Uri.parse('$baseUrl/create'),
          headers: <String, String>{
            'Content-Type': 'application/json; charset=UTF-8',
          },
          body: jsonEncode(<String, String>{
            'username': username,
            'email': email,
            'password': password
          }),
        );

        if(response.statusCode == 200) {
            var data = jsonDecode(response.body);
            user = User.fromJson(data);  
            return true;
        } else {
            return false;
        }
  }

  Future<bool> authenticateUser(String username, String password) async {
        //https://pub.dev/packages/shared_preferences
        final SharedPreferences prefs = await SharedPreferences.getInstance();

        final response = await http.post(
          Uri.parse('$baseUrl/users/auth'),
          headers: <String, String>{
            'Content-Type': 'application/json; charset=UTF-8',
          },
          body: jsonEncode(<String, String>{
            'username': username,
            'password': password
          }),
        );

              // Decode the response body
        var data = jsonDecode(response.body);

        // Make sure you access 'jwt' properly from the decoded data
        if (data.containsKey('jwt')) {
          await prefs.setString('jwt', data['jwt'].toString());
          final String? token = prefs.getString('jwt');
          return true;
        } else {
          return false;
        }
  }

  Future<bool> getUser(String id) async {
    final response = await http.get(Uri.parse('$baseUrl/$id'));

    if (response.statusCode == 200) {
      var data = jsonDecode(response.body);
            user = User.fromJson(data);  
            return true;
    } else {
      return false;
    }
  }

  Future<bool> getAllUsers() async {
    final response = await http.get(Uri.parse('$baseUrl/'));

    if (response.statusCode == 200) {
      var data = jsonDecode(response.body) as List;
      users = data.map((userJson) => User.fromJson(userJson)).toList();
            return true;
    } else {
      return false;
    }
  }

  // Update (PUT)
  Future<bool> updateUser(User user) async {
    final response = await http.put(
      Uri.parse('$baseUrl/$user.id'),
      headers: {
        'Content-Type': 'application/json',
      },
      body: jsonEncode({
        'username': user.username,
        'email': user.email,
      }),
    );

    if (response.statusCode == 200) {
      return true;
    } else {
      return false;
    }
  }

  // Delete (DELETE)
  Future<bool> deleteUser(String id) async {
    final response = await http.delete(Uri.parse('$baseUrl/$id'));

    if (response.statusCode == 200) {
      return true;
    } else {
      return false;
    }
  }

}

File: react/react-animated-barnacle/src/Users/UserService.js
import User from './UserModel';

class UserService {
    constructor() {
      this.baseUrl = "http://127.0.0.1:8080/users";
      this.user = null; 
      this.users = [];  
    }
  
    showErrorDialog(message) {
      alert(`Error: ${message}`);
    }
  
    showSuccessDialog(message) {
      alert(`Success: ${message}`);
    }
  
    async createUser(username, email, password) {
      try {
        const response = await fetch(`${this.baseUrl}/create`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json; charset=UTF-8',
          },
          body: JSON.stringify({ username, email, password }),
        });
  
        if (response.ok) {
          const data = await response.json();
          this.user = User.fromJson(data);
          this.showSuccessDialog("User created successfully!");
          return true;
        } else {
          this.showErrorDialog("Failed to create user.");
          return false;
        }
      } catch (error) {
        this.showErrorDialog(`Error: ${error.message}`);
        return false;
      }
    }
  
    async authenticateUser(username, password) {
      try {
        const response = await fetch(`${this.baseUrl}/auth`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json; charset=UTF-8',
          },
          body: JSON.stringify({ username, password }),
        });
  
        const data = await response.json();
  
        // Assuming 'jwt' is included in the response
        if (data.jwt) {
          localStorage.setItem('jwt', data.jwt); // Store token in local storage
          return true;
        } else {
          this.showErrorDialog("Authentication failed.");
          return false;
        }
      } catch (error) {
        this.showErrorDialog(`Error: ${error.message}`);
        return false;
      }
    }
  
    async getUser(id) {
      try {
        const response = await fetch(`${this.baseUrl}/${id}`);
  
        if (response.ok) {
          const data = await response.json();
          this.user = User.fromJson(data);
          return true;
        } else {
          this.showErrorDialog("Failed to fetch user.");
          return false;
        }
      } catch (error) {
        this.showErrorDialog(`Error: ${error.message}`);
        return false;
      }
    }
  
    async getAllUsers() {
      try {
        const response = await fetch(this.baseUrl);
  
        if (response.ok) {
          const data = await response.json();
          this.users = data.map(userJson => User.fromJson(userJson));
          return true;
        } else {
          this.showErrorDialog("Failed to fetch users.");
          return false;
        }
      } catch (error) {
        this.showErrorDialog(`Error: ${error.message}`);
        return false;
      }
    }
  
    async updateUser(user) {
      try {
        const response = await fetch(`${this.baseUrl}/${user.id}`, {
          method: 'PUT',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            username: user.username,
            email: user.email,
          }),
        });
  
        if (response.ok) {
          this.showSuccessDialog("User updated successfully!");
          return true;
        } else {
          this.showErrorDialog("Failed to update user.");
          return false;
        }
      } catch (error) {
        this.showErrorDialog(`Error: ${error.message}`);
        return false;
      }
    }
  
    async deleteUser(id) {
      try {
        const response = await fetch(`${this.baseUrl}/${id}`, {
          method: 'DELETE',
        });
  
        if (response.ok) {
          this.showSuccessDialog("User deleted successfully!");
          return true;
        } else {
          this.showErrorDialog("Failed to delete user.");
          return false;
        }
      } catch (error) {
        this.showErrorDialog(`Error: ${error.message}`);
        return false;
      }
    }
  }

  export default UserService;